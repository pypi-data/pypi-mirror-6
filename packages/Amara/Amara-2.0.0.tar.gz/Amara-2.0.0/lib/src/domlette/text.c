#define PY_SSIZE_T_CLEAN
#include "domlette_interface.h"

/** Private Routines **************************************************/

/* nothing to see here */

/** Public C API *******************************************************/

/* Remove the preprocessor macros so that we can export the entry points. */

#undef Text_New
TextObject *Text_New(PyObject *data)
{
  return CharacterData_New(TextObject, &DomletteText_Type, data);
}

/** Python Methods ****************************************************/

static PyMethodDef text_methods[] = {
  { NULL }
};

/** Python Members *****************************************************/

static PyMethodDef text_members[] = {
  { NULL }
};

/** Python Computed Members ********************************************/

static PyGetSetDef text_getset[] = {
  { NULL }
};

/** Type Object ********************************************************/

static char text_doc[] = "\
text(data) -> text object\n\
\n\
The `text` interface represents the contents of a text node.";

PyTypeObject DomletteText_Type = {
  /* PyObject_HEAD     */ PyObject_HEAD_INIT(NULL)
  /* ob_size           */ 0,
  /* tp_name           */ Domlette_MODULE_NAME "." "text",
  /* tp_basicsize      */ sizeof(TextObject),
  /* tp_itemsize       */ 0,
  /* tp_dealloc        */ (destructor) 0,
  /* tp_print          */ (printfunc) 0,
  /* tp_getattr        */ (getattrfunc) 0,
  /* tp_setattr        */ (setattrfunc) 0,
  /* tp_compare        */ (cmpfunc) 0,
  /* tp_repr           */ (reprfunc) 0,
  /* tp_as_number      */ (PyNumberMethods *) 0,
  /* tp_as_sequence    */ (PySequenceMethods *) 0,
  /* tp_as_mapping     */ (PyMappingMethods *) 0,
  /* tp_hash           */ (hashfunc) 0,
  /* tp_call           */ (ternaryfunc) 0,
  /* tp_str            */ (reprfunc) 0,
  /* tp_getattro       */ (getattrofunc) 0,
  /* tp_setattro       */ (setattrofunc) 0,
  /* tp_as_buffer      */ (PyBufferProcs *) 0,
  /* tp_flags          */ (Py_TPFLAGS_DEFAULT |
                           Py_TPFLAGS_BASETYPE),
  /* tp_doc            */ (char *) text_doc,
  /* tp_traverse       */ (traverseproc) 0,
  /* tp_clear          */ (inquiry) 0,
  /* tp_richcompare    */ (richcmpfunc) 0,
  /* tp_weaklistoffset */ 0,
  /* tp_iter           */ (getiterfunc) 0,
  /* tp_iternext       */ (iternextfunc) 0,
  /* tp_methods        */ (PyMethodDef *) text_methods,
  /* tp_members        */ (PyMemberDef *) text_members,
  /* tp_getset         */ (PyGetSetDef *) text_getset,
  /* tp_base           */ (PyTypeObject *) 0,
  /* tp_dict           */ (PyObject *) 0,
  /* tp_descr_get      */ (descrgetfunc) 0,
  /* tp_descr_set      */ (descrsetfunc) 0,
  /* tp_dictoffset     */ 0,
  /* tp_init           */ (initproc) 0,
  /* tp_alloc          */ (allocfunc) 0,
  /* tp_new            */ (newfunc) 0,
  /* tp_free           */ 0,
};

/** Module Interface **************************************************/

int DomletteText_Init(PyObject *module)
{
  PyObject *dict, *value;

  DomletteText_Type.tp_base = &DomletteCharacterData_Type;
  if (PyType_Ready(&DomletteText_Type) < 0)
    return -1;

  dict = DomletteText_Type.tp_dict;

  value = PyString_FromString("text");
  if (value == NULL)
    return -1;
  if (PyDict_SetItemString(dict, "xml_type", value))
    return -1;
  Py_DECREF(value);
  /* add the "typecode" character for use with `xml_nodeid` */
  value = PyString_FromString("t");
  if (value == NULL)
    return -1;
  if (PyDict_SetItemString(dict, "xml_typecode", value) < 0)
    return -1;
  Py_DECREF(value);

  if (PyDict_SetItemString(dict, "xsltOutputEscaping", Py_True))
    return -1;

  Py_INCREF(&DomletteText_Type);
  return PyModule_AddObject(module, "text", (PyObject*)&DomletteText_Type);
}

void DomletteText_Fini(void)
{
  PyType_CLEAR(&DomletteText_Type);
}
