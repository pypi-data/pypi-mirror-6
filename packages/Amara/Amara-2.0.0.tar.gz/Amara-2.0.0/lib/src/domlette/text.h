#ifndef DOMLETTE_TEXT_H
#define DOMLETTE_TEXT_H

#ifdef __cplusplus
extern "C" {
#endif

#include "Python.h"
#include "characterdata.h"

  typedef CharacterDataObject TextObject;

#define Text(op) ((TextObject *)(op))
#define Text_GET_VALUE CharacterData_GET_VALUE

#ifdef Domlette_BUILDING_MODULE

  extern PyTypeObject DomletteText_Type;

#define Text_Check(op) PyObject_TypeCheck((op), &DomletteText_Type)
#define Text_CheckExact(op) ((op)->ob_type == &DomletteText_Type)
#define Text_SET_VALUE CharacterData_SET_VALUE

  /* Module Methods */
  int DomletteText_Init(PyObject *module);
  void DomletteText_Fini(void);

  /* Text Methods */
  TextObject *Text_New(PyObject *data);
#define Text_New(data) \
  CharacterData_New(TextObject, &DomletteText_Type, (data))

#endif /* Domlette_BUILDING_MODULE */

#ifdef __cplusplus
}
#endif

#endif /* DOMLETTE_TEXT_H */
