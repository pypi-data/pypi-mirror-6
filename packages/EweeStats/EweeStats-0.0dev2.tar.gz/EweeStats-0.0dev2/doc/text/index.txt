
Documentation de EweeStats
**************************

EweeStats est un programme d'analyse et de traitement des données
destiné à l'Ewee. Il lit les données en analogique, les traite et les
présente. EweeStats démarre au lancement de la RPi et enregistre
toutes les données tant que l'Ewee fonctionne.


Sections
========

* Choix matériels et logiciels

* Lecture et transmission des mesures

* Traitement et enregistrement des données

* Présentation des données


Choix matériels
---------------

* Choix matériels et logiciels

  * Matériel utilisé

  * Logiciels


Acquisition et transmission des mesures
---------------------------------------

* Lecture et transmission des mesures

  * Transmission côté Arduino

  * Transmission côté Raspberry Pi


Traitement des données
----------------------

* Traitement et enregistrement des données

  * Organisation des données

  * Enregistrement

  * Formatage des données pour l'outil graphique


Présentation des mesures
------------------------

* Présentation des données

  * Support d'affichage

  * Diffusion des données


Documentation de conception
===========================

* Liste des modules utilisés :

* Architecture du programme

* Initialisation du programme

  * Initialisation de l'écran LCD

  * Initialisation de l'Arduino

  * Initialisation des fichiers de sauvegarde

    * Chemin des fichiers

    * Création des fichiers

  * Initialisation de l'horodatage

* Générisation des opérations

* Lecture/écriture et horodatage des données

  * Lecture

  * Horodatage

  * Ecriture

* IHM / IMH

  * Interface homme machine

  * Interface machine homme

* Création du graphique

  * Lecture des données

  * Formatage des données

  * Paramétrage du graphique

* Multithreading

  * Synchronisation des threads

* Diffusion du graphique


Indices and tables
==================

* *Index*

* *Index du module*

* *Page de recherche*
