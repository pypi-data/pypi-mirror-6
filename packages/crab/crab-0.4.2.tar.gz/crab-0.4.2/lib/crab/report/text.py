# Copyright (C) 2012 Science and Technology Facilities Council.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from crab import CrabEvent, CrabStatus

def report_to_text(report, event_list=True):
    lines = []
    sections = ['error', 'warning', 'ok']
    titles = ['Jobs with Errors', 'Jobs with Warnings', 'Successful Jobs']

    for (section, title) in zip(sections, titles):
        jobs = getattr(report, section)
        if jobs:
            lines.append(title)
            lines.append('=' * len(title))
            lines.append('')
            for id_ in jobs:
                lines.append('    ' + _summary_line(report, id_))
            lines.append('')

    if event_list:
        lines.append('Event Listing')
        lines.append('=============')
        lines.append('')

        for id_ in set.union(report.error, report.warning, report.ok):
            subhead = _summary_line(report, id_)
            lines.append(subhead)
            lines.append('-' * len(subhead))
            lines.append('')

            for e in report.events[id_]:
                lines.append('    ' + _event_line(e))

                if e['type'] == CrabEvent.FINISH:
                    finishid = e['eventid']
                    if finishid in report.stdout and report.stdout[finishid]:
                        lines.extend(_output_lines(8, 'Std. Out.',
                                                   report.stdout[finishid]))
                    if finishid in report.stderr and report.stderr[finishid]:
                        lines.extend(_output_lines(8, 'Std. Error',
                                                   report.stderr[finishid]))

            lines.append('')

    return "\n".join(lines)


def _summary_line(report, id_):
    info = report.info[id_]
    return '{0:10} {1:10} {2}'.format(info['host'], info['user'], info['title'])

def _event_line(event):
    return '{0:10} {1:10} {2}'.format(CrabEvent.get_name(event['type']),
                                CrabStatus.get_name(event['status']),
                                event['datetime'])

def _output_lines(indent, title, text):
    lines = []
    for line in text.strip().split('\n'):
        if lines:
            head = ''
        else:
            head = title
        lines.append('{0}{1:10} {2}'.format(' ' * indent, head, line))

    return lines
