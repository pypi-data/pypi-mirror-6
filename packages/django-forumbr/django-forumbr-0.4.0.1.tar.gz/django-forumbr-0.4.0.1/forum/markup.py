# -*- coding:utf-8 -*-

"""
Only bbcode is supported so far.
"""

from .constants import BBCODE, MARKDOWN, TEXTILE
from .constants import RST, PLAIN_HTML, PLAIN_TEXT

# FIXME: setup logging

try:
    import textile
except ImportError:
    textile = None
    print 'textile module not found. Textile support is off.'

try:
    import markdown
except ImportError:
    markdown = None
    print 'markdown module not found. Markdown support is off.'

try:
    import bbcode
except ImportError:
    bbcode = None
    print 'bbcode module not found. BBCode support is off.'

try:
    from docutils.core import publish_parts
except ImportError:
    publish_parts = None
    print 'docutils module not found. RST support is off.'

try:
    import html5lib
    from html5lib import sanitizer

    html_sanitizer = html5lib.HTMLParser(tokenizer=sanitizer.HTMLSanitizer)
except ImportError:
    html_sanitizer = None
    print 'html5lib not found. Plain html support is off.'


def render_markup(markup, text):
    if text is None:
        return None

    if markup == BBCODE:
        if bbcode is not None:
            return bbcode.render_html(text)
    elif markup == MARKDOWN:
        if markdown is not None:
            return markdown.markdown(text)
    elif markup == TEXTILE:
        if textile is not None:
            return textile.textile(text)
    elif markup == RST:
        if publish_parts is not None:
            return publish_parts(text, writer_name='html')['fragment']
    elif markup == PLAIN_HTML:
        if html_sanitizer is not None:
            return html_sanitizer.parse(text)
    return text