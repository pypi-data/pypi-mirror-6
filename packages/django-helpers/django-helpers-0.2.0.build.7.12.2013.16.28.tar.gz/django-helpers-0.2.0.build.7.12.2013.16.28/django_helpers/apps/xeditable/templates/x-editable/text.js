$.dj.xEditable("#" + '{{ field_id }}', "#" + '{{ field_id }}', '{{ save_url }}', {
    type : 'text',
    pk : '{{ pk }}',
    name : '{{ name }}',
    placement : 'top',
    title : '{{ title }}'
});