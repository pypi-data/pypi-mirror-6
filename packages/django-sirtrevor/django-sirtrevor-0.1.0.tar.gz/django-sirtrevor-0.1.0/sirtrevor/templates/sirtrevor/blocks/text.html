{% load markdown2 %}
<div class="content-block text-block">
    {{ text|markdown }}
</div>
