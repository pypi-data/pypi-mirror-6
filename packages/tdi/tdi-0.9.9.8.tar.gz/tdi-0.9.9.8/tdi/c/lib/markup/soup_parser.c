/*
 * Copyright 2013
 * Andr\xe9 Malo or his licensors, as applicable
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "cext.h"
#include "tdi.h"

#include "tdi_soup_lexer.h"
#include "tdi_soup_parser.h"

/* -------------------------- BEGIN DECLARATIONS ------------------------- */

typedef struct tagstack {
    struct tagstack *next;
    PyObject *normname;
    PyObject *name;
} tagstack;

struct tdi_soup_parser {
    /* Last error, to be evaluated by the caller */
    tdi_parser_error last_error;

    /* Lexer */
    tdi_soup_lexer *lexer;

    /* Event receiver */
    tdi_parser_callback *cb;
    void *cb_ctx;

    /* Nestable cb */
    tdi_soup_parser_nestable_callback *nestable;
    void *nestable_ctx;

    /* cdata cb */
    tdi_soup_parser_cdata_callback *cdata;
    void *cdata_ctx;

    /* empty cb */
    tdi_soup_parser_empty_callback *empty;
    void *empty_ctx;

    /* normalize cb */
    tdi_soup_parser_normalize_callback *normalize;
    void *normalize_ctx;

    /* Parser state stuff */
    tagstack *tagstack;
    PyObject *lastopen;
    int inempty;
};

/* --------------------------- END DECLARATIONS -------------------------- */

/* ---------------------- BEGIN tagstack DEFINITION ---------------------- */

static int
tagstack_push(tagstack **stack_, PyObject *normname, PyObject *name)
{
    tagstack *stack;

    if (!(stack = PyMem_Malloc(sizeof *stack)))
        return -1;

    stack->normname = (Py_INCREF(normname), normname);
    stack->name = (Py_INCREF(name), name);
    stack->next = *stack_;
    *stack_ = stack;

    return 0;
}

static void
tagstack_pop(tagstack **stack_)
{
    tagstack *tmp, *stack = *stack_;
    if (stack) {
        *stack_ = tmp = stack->next;
        Py_DECREF(stack->normname);
        Py_DECREF(stack->name);
        PyMem_Free(stack);
    }
}

/* ----------------------- END tagstack DEFINITION ----------------------- */

static int
lexer_error(tdi_soup_parser *self)
{
    switch (tdi_soup_lexer_error_get(self->lexer)) {
    case TDI_LEXER_ERR_ENV:
        break;

    case TDI_LEXER_ERR_EOF:
        self->last_error = TDI_PARSER_ERR_LEXER_EOF;
        break;

    case TDI_LEXER_ERR_FINAL:
        self->last_error = TDI_PARSER_ERR_LEXER_FINAL;
        break;
    }

    return -1;
}

static int
close_empty(tdi_soup_parser *self)
{
    PyObject *data, *name = self->tagstack->name;
    tdi_parser_event event;
    int res;

    if (!(data = PyString_FromString("")))
        return -1;
    Py_INCREF(name);

    self->inempty = 0;
    tagstack_pop(&self->tagstack);

    event.type = TDI_PARSER_EVENT_ENDTAG;
    event.info.endtag.name = name;
    event.info.endtag.data = data;
    res = !self->cb(&event, self->cb_ctx) ? 0 : -1;
    Py_DECREF(name);
    Py_DECREF(data);

    return res;
}

static int
handle_starttag(tdi_soup_parser *self, tdi_parser_event *event,
                tdi_lexer_event *event_)
{
    PyObject *name, *normname, *tmp, *data;
    int res;

    if (self->inempty && close_empty(self) == -1) return -1;

    /* sanitize */
    if (PyString_GET_SIZE(event_->info.starttag.name) == 0
        && PyList_GET_SIZE(event_->info.starttag.attr) == 0) {
        name = (Py_INCREF(self->lastopen), self->lastopen);
    }
    else {
        name = event_->info.starttag.name;
        Py_INCREF(name);
        Py_CLEAR(self->lastopen);
        self->lastopen = (Py_INCREF(name), name);
    }

    if (!(normname = self->normalize(self->normalize_ctx, name)))
        goto error;

    /* close unnestables */
    while (self->tagstack) {
        res = self->nestable(self->nestable_ctx,
                             self->tagstack->normname, normname);
        if (res == -1)
            goto error_normname;
        if (res) break;

        event->type = TDI_PARSER_EVENT_ENDTAG;
        if (!(data = PyString_FromString("")))
            goto error_normname;
        tmp = self->tagstack->name;
        event->info.endtag.name = (Py_INCREF(tmp), tmp);
        event->info.endtag.data = data;
        tagstack_pop(&self->tagstack);
        res = !self->cb(event, self->cb_ctx) ? 0 : -1;
        Py_DECREF(tmp);
        Py_DECREF(data);
        if (res == -1)
            goto error_normname;
    }

    /* CDATA */
    if (!event_->info.starttag.closed) {
        if ((res = self->cdata(self->cdata_ctx, normname)) == -1)
            goto error_normname;
        if (res) {
            res = tdi_soup_lexer_state_cdata(self->lexer, self->normalize,
                                             self->normalize_ctx, normname);
            if (res == -1) {
                lexer_error(self);
                goto error_normname;
            }
        }
    }

    /* pass event */
    event->type = TDI_PARSER_EVENT_STARTTAG;
    event->info.starttag.name = name;
    event->info.starttag.attr = event_->info.starttag.attr;
    event->info.starttag.closed = event_->info.starttag.closed;
    event->info.starttag.data = event_->info.starttag.data;
    if (self->cb(event, self->cb_ctx))
        goto error_normname;

    /* Maintain stack */
    if (!event_->info.starttag.closed) {
        if (tagstack_push(&self->tagstack, normname, name) == -1)
            goto error_normname;
        if ((res = self->empty(self->empty_ctx, normname)) == -1)
            goto error_normname;
        if (res)
            self->inempty = 1;
    }

    /* cleanup & finish */
    Py_DECREF(normname);
    Py_DECREF(name);
    return 0;

error_normname:
    Py_DECREF(normname);
error:
    Py_DECREF(name);

    if (!self->last_error)
        self->last_error = TDI_PARSER_ERR_ENV;
    return -1;
}

static int
handle_endtag(tdi_soup_parser *self, tdi_parser_event *event,
              tdi_lexer_event *event_)
{
    PyObject *normname, *name = event_->info.endtag.name;
    PyObject *toclose, *original, *data;
    tagstack *item;
    int cmp;

    if (!self->tagstack) {
        Py_INCREF(name);
    }
    else {
        if (PyString_GET_SIZE(name) == 0)
            name = self->tagstack->name;
        Py_INCREF(name);
        if (!(normname = self->normalize(self->normalize_ctx, name)))
            goto error;

        /* look if we're closing anything open.
         * This also resolves misnestings.
         */
        for (item = self->tagstack; item; item = item->next) {
            if (PyObject_Cmp(item->normname, normname, &cmp) == -1)
                goto error_normname;
            if (!cmp) break;
        }
        /* Found */
        if (item) {
            self->inempty = 0;
            while (self->tagstack) {
                toclose = self->tagstack->normname; Py_INCREF(toclose);
                original = self->tagstack->name; Py_INCREF(original);
                tagstack_pop(&self->tagstack);

                if (PyObject_Cmp(toclose, normname, &cmp) == -1)
                    goto error_loop;
                if (!cmp) {
                    Py_DECREF(original);
                    Py_DECREF(toclose);
                    break;
                }
                if (!(data = PyString_FromString("")))
                    goto error_loop;

                event->type = TDI_PARSER_EVENT_ENDTAG;
                event->info.endtag.name = original;
                event->info.endtag.data = data;
                if (self->cb(event, self->cb_ctx))
                    goto error_data;

                Py_DECREF(data);
                Py_DECREF(original);
                Py_DECREF(toclose);
            }
        }
        Py_DECREF(normname);
    }

    if (self->inempty && close_empty(self) == -1) goto error;

    event->type = TDI_PARSER_EVENT_ENDTAG;
    event->info.endtag.name = name;
    event->info.endtag.data = event_->info.endtag.data;
    if (self->cb(event, self->cb_ctx))
        goto error;

    Py_DECREF(name);
    return 0;

error_data:
    Py_DECREF(data);
error_loop:
    Py_DECREF(original);
    Py_DECREF(toclose);
error_normname:
    Py_DECREF(normname);
error:
    Py_DECREF(name);

    if (!self->last_error)
        self->last_error = TDI_PARSER_ERR_ENV;
    return -1;
}


static int
lexer_callback(tdi_lexer_event *event_, void *self_)
{
    tdi_soup_parser *self = self_;
    tdi_parser_event event;

    switch (event_->type) {
    case TDI_LEXER_EVENT_STARTTAG:
        return handle_starttag(self, &event, event_);

    case TDI_LEXER_EVENT_ENDTAG:
        return handle_endtag(self, &event, event_);

    case TDI_LEXER_EVENT_TEXT:
        if (self->inempty && close_empty(self) == -1) return -1;

        event.type = TDI_PARSER_EVENT_TEXT;
        event.info.text.data = event_->info.text.data;
        return !self->cb(&event, self->cb_ctx) ? 0 : -1;

    case TDI_LEXER_EVENT_COMMENT:
        if (self->inempty && close_empty(self) == -1) return -1;

        event.type = TDI_PARSER_EVENT_COMMENT;
        event.info.comment.data = event_->info.comment.data;
        return !self->cb(&event, self->cb_ctx) ? 0 : -1;

    case TDI_LEXER_EVENT_MSECTION:
        if (self->inempty && close_empty(self) == -1) return -1;

        event.type = TDI_PARSER_EVENT_MSECTION;
        event.info.msection.data = event_->info.msection.data;
        event.info.msection.name = event_->info.msection.name;
        event.info.msection.value = event_->info.msection.value;
        return !self->cb(&event, self->cb_ctx) ? 0 : -1;

    case TDI_LEXER_EVENT_DECL:
        if (self->inempty && close_empty(self) == -1) return -1;

        event.type = TDI_PARSER_EVENT_DECL;
        event.info.decl.data = event_->info.decl.data;
        event.info.decl.name = event_->info.decl.name;
        event.info.decl.value = event_->info.decl.value;
        return !self->cb(&event, self->cb_ctx) ? 0 : -1;

    case TDI_LEXER_EVENT_PI:
        if (self->inempty && close_empty(self) == -1) return -1;

        event.type = TDI_PARSER_EVENT_PI;
        event.info.pi.data = event_->info.pi.data;
        return !self->cb(&event, self->cb_ctx) ? 0 : -1;

    case TDI_LEXER_EVENT_ESCAPE:
        break;
    }

    /* Should not happen */
    PyErr_SetNone(PyExc_AssertionError);
    self->last_error = TDI_PARSER_ERR_ENV;
    return -1;
}

/* ------------------------------ BEGIN API ------------------------------ */

tdi_soup_parser *
tdi_soup_parser_new(tdi_parser_callback *cb, void *cb_ctx,
                    tdi_soup_parser_nestable_callback *nestable,
                    void *nestable_ctx,
                    tdi_soup_parser_cdata_callback *cdata,
                    void *cdata_ctx,
                    tdi_soup_parser_empty_callback *empty,
                    void *empty_ctx,
                    tdi_soup_parser_normalize_callback *normalize,
                    void *normalize_ctx)
{
    tdi_soup_parser *self;

    if (!(self = PyMem_Malloc(sizeof *self)))
        return NULL;

    if (!(self->lexer = tdi_soup_lexer_new(lexer_callback, self, 1)))
        goto error;
    if (!(self->lastopen = PyString_FromString("")))
        goto error_lexer;

    self->tagstack = NULL;
    self->inempty = 0;

    self->last_error = 0;
    self->cb = cb;
    self->cb_ctx = cb_ctx;
    self->nestable = nestable;
    self->nestable_ctx = nestable_ctx;
    self->cdata = cdata;
    self->cdata_ctx = cdata_ctx;
    self->empty = empty;
    self->empty_ctx = empty_ctx;
    self->normalize = normalize;
    self->normalize_ctx = normalize_ctx;

    return self;

error_lexer:
    TDI_SOUP_LEXER_CLEAR(self->lexer);
error:
    PyMem_Free(self);
    return NULL;
}

void
tdi_soup_parser_del(tdi_soup_parser *self)
{
    if (self) {
        TDI_SOUP_LEXER_CLEAR(self->lexer);
        Py_CLEAR(self->lastopen);
        while (self->tagstack)
            tagstack_pop(&self->tagstack);
        PyMem_Free(self);
    }
}

int
tdi_soup_parser_feed(tdi_soup_parser *self, PyObject *food)
{
    if (tdi_soup_lexer_feed(self->lexer, food) == -1)
        return lexer_error(self);

    return 0;
}

int
tdi_soup_parser_finalize(tdi_soup_parser *self)
{
    PyObject *data, *name;
    tdi_parser_event event;
    int res;

    if (tdi_soup_lexer_finalize(self->lexer) == -1)
        return lexer_error(self);

    while (self->tagstack) {
        if (!(data = PyString_FromString(""))) {
            self->last_error = TDI_PARSER_ERR_ENV;
            return -1;
        }
        name = self->tagstack->name;
        Py_INCREF(name);
        tagstack_pop(&self->tagstack);

        event.type = TDI_PARSER_EVENT_ENDTAG;
        event.info.endtag.name = name;
        event.info.endtag.data = data;
        res = self->cb(&event, self->cb_ctx);
        Py_DECREF(name);
        Py_DECREF(data);
        if (res) {
            self->last_error = TDI_PARSER_ERR_ENV;
            return -1;
        }
    }

    return 0;
}

const char *
tdi_soup_parser_lexer_state_get(tdi_soup_parser *self)
{
    return tdi_soup_lexer_state_get(self->lexer);
}

tdi_parser_error
tdi_soup_parser_error_get(tdi_soup_parser *self)
{
    return self->last_error;
}

/* ------------------------------- END API ------------------------------- */

