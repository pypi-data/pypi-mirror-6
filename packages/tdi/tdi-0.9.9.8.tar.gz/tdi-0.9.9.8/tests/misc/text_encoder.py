#!/usr/bin/env python
import warnings as _warnings
_warnings.resetwarnings()
_warnings.filterwarnings('error')

from tdi.markup.text import encoder as _encoder


encoder = _encoder.TextEncoder('utf-8')

print encoder.starttag('a', [('b', 'c'), ('d', None)], True)
print encoder.starttag('a', [('b1', None), ('c2', 'd3')], False)
print encoder.endtag('ff')
print encoder.name(u'ggg')
print encoder.name('ggg')
print encoder.attribute(u'\u20ac"')
print encoder.attribute(u'\u20ac')
print encoder.attribute(u'\u20ac"'.encode('utf-8'))
print encoder.attribute(u'\u20ac'.encode('utf-8'))
print encoder.content(u'\u20ac')
print encoder.content(u'\u20ac'.encode('utf-8'))
print encoder.encode(u'\u20ac')
print encoder.escape('lalala')
print encoder.escape('[lalala]')
