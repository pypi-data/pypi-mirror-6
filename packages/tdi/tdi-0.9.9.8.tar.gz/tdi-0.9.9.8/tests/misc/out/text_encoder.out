[[a b=c d]]
[a b1 c2=d3]
[/ff]
ggg
ggg
"€\""
"€"
"€\""
"€"
€
€
€
lalala
[]lalala]
